File: Programmers/coding_basic_training/day12/181893.java
package Programmers.coding_basic_training.day12;

/*
 * 배열 조각하기
 * https://school.programmers.co.kr/learn/courses/30/lessons/181893
*/

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

class _181893 {
    public int[] solution(int[] arr, int[] query) {
        List<Integer> arrList = Arrays.stream(arr)
                .boxed()
                .collect(Collectors.toList());

        for (int i = 0; i < query.length; i++) {
            if (i % 2 == 0) {
                arrList = arrList.subList(0, query[i] + 1);
            } else {
                arrList = arrList.subList(query[i], arrList.size());
            }
        }

        return arrList.stream()
                .mapToInt(Integer::intValue)
                .toArray();
    }

    // optimized solution
    // 필요한 인덱스를 찾아서 배열을 한 번만 자르는 방식
    public int[] solutionOpt(int[] arr, int[] query) {
        int startIndex = 0;
        int endIndex = arr.length - 1;

        for (int queryIndex = 0; queryIndex < query.length; queryIndex++) {
            int cutIndex = query[queryIndex];

            if (queryIndex % 2 == 0) {
                endIndex = startIndex + cutIndex;
            } else {
                startIndex += cutIndex;
            }
        }

        return Arrays.copyOfRange(arr, startIndex, endIndex + 1);
    }
}

File: Programmers/coding_basic_training/day12/181893.cpp

/*
 * 배열 조각하기
 * https://school.programmers.co.kr/learn/courses/30/lessons/181893
*/

#include <vector>

using namespace std;

vector<int> solution(vector<int> arr, vector<int> query) {
    for (int i = 0; i < query.size(); i++) {
        if (i % 2 == 0) {
            arr.erase(arr.begin() + query[i] + 1, arr.end());
        } else {
            arr.erase(arr.begin(), arr.begin() + query[i]);
        }
    }

    return arr;
}

// optimized solution
vector<int> solutionOpt(const vector<int>& arr, const vector<int>& query) {
    int startIndex = 0;
    int endIndex = static_cast<int>(arr.size()) - 1;

    for (int queryIndex = 0; queryIndex < static_cast<int>(query.size()); ++queryIndex) {
        int cutIndex = query[queryIndex];

        if (queryIndex % 2 == 0) {
            endIndex = startIndex + cutIndex;
        } else {
            startIndex += cutIndex;
        }
    }
    
    return vector<int>(
        arr.begin() + startIndex,
        arr.begin() + endIndex + 1
    );
}